internal/c/libqb/src/filesystem.cpp

```diff
@@ -0,0 +1,986 @@
+//-----------------------------------------------------------------------------------------------------
+//  QB64-PE filesystem related functions
+//-----------------------------------------------------------------------------------------------------
+
+#include "libqb-common.h"
+
+#include "filepath.h"
+#include "filesystem.h"
+
+#include "../../libqb.h"
+
+#include <algorithm>
+#include <dirent.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#ifdef QB64_WINDOWS
+#    include <shlobj.h>
+#else
+#    include <sys/statvfs.h>
+#endif
+
+#ifdef QB64_BACKSLASH_FILESYSTEM
+#    define FS_PATH_SEPARATOR '\\'
+#else
+#    define FS_PATH_SEPARATOR '/'
+#endif
+
+#if (FILENAME_MAX > 4096)
+#    define FS_PATHNAME_LENGTH_MAX FILENAME_MAX
+#else
+#    define FS_PATHNAME_LENGTH_MAX 4096
+#endif
+
+/// @brief This is a global variable that is set on startup and holds the directory that was current when the program was loaded
+qbs *g_startDir = nullptr;
+
+/// @brief Gets the current working directory
+/// @return A qbs containing the current working directory or an empty string on error
+qbs *func__cwd() {
+    std::string path;
+    qbs *final;
+
+    path.resize(FILENAME_MAX, '\0');
+
+    for (;;) {
+        if (getcwd(&path[0], path.size())) {
+            auto size = strlen(path.c_str());
+            final = qbs_new(size, 1);
+            memcpy(final->chr, &path[0], size);
+
+            return final;
+        } else {
+            if (errno == ERANGE)
+                path.resize(path.size() << 1, '\0'); // buffer size was not sufficient; try again with a larger buffer
+            else
+                break; // some other error occurred
+        }
+    }
+
+    final = qbs_new(0, 1);
+    error(QB_ERROR_INTERNAL_ERROR);
+
+    return final;
+}
+
+/// @brief Returns true if the specified directory exists
+/// @param path The directory to check for
+/// @return True if the directory exists
+static inline bool FS_DirectoryExists(const char *path) {
+#ifdef QB64_WINDOWS
+    auto x = GetFileAttributesA(path);
+
+    return x != INVALID_FILE_ATTRIBUTES && (x & FILE_ATTRIBUTE_DIRECTORY);
+#else
+    struct stat info;
+
+    return stat(path, &info) == 0 && S_ISDIR(info.st_mode);
+#endif
+}
+
+/// @brief Known directories (primarily Windows based, but we'll do our best to emulate on other platforms)
+enum class FS_KnownDirectory {
+    HOME = 0,
+    DESKTOP,
+    DOCUMENTS,
+    PICTURES,
+    MUSIC,
+    VIDEOS,
+    DOWNLOAD,
+    APP_DATA,
+    LOCAL_APP_DATA,
+    PROGRAM_DATA,
+    SYSTEM_FONTS,
+    USER_FONTS,
+    TEMP,
+    PROGRAM_FILES,
+    PROGRAM_FILES_32,
+};
+
+/// @brief This populates path with the full path for a known directory
+/// @param kD Is a value from FS_KnownDirectory (above)
+/// @return Is the string that will receive the directory path. The string may be changed
+static std::string FS_GetKnownDirectory(FS_KnownDirectory kD) {
```

Comment:
You might get something a bit cleaner if you have two separate `FS_GetKnownDirectory()` implementations for Windows Vs Unix. There would be some slight duplication in the structure, but a lot less `#ifdef`s to read.

Reply:
That's a valid callout. It does look a little messy. I'll clean up that part.

Reply:
This is done.